# Metagenomics Data Processing
Welcome to the ADOR lab Metagenomics Data Processing github repo!
This is aimed to be a resource for future metagenomics data processing and analyses.

Information covering the following topics is included: 

1. setting_up
	- Setting up a text edittor and git/github.
	- Designing a study
2. command_line
	- Command line essentials, tips and tricks.
3. computer_cluster
	- Working on a computer cluster basics
	- Using CU Boulder's Alpine (very similar to Peta Library)
	- Explanation of snakemake pipelines
4. transferring_data
	- scp, wget, sftp
	- Globus 
5. QA_and_QC on metagenomic data.
	- How to perform quality control checks on metagenomic sequences using:
		- Fastqc
		- Multiqc
	- Removing low quality sequence reads with:
		- Trimmomatic
	- Explore sequencing depth in R
		- Setting thresholds for metagenomic processing.
		- Sequencing and the needed coverage.
	- Remove human reads from microbial reads using:
		- Hostile
	- Performing all of the above on a cluster environment
6. toolbox
	- A place for tutorials and scripts on the tools used on the metagenomic data after QA and QC. This includes:
		- Humann3 (& Metaphlan)
		- Kraken2
		- PhyloFlash

new:
![image](https://github.com/UC-Boulder/metagenomics_data_processing/assets/104112036/80d19010-5460-449f-b953-ffbd438770b4)

old:
![image](https://github.com/UC-Boulder/metagenomics_data_processing/assets/104112036/3ec99367-9e03-4fe9-8a71-d66717a88d29)


