
# QA and QC on Alpine

It's super exciting to get new metagenomic sequencing data to work with!
While you might want to jump right into your research questions, the
first thing you need to do is assess the quality of the sequencing data,
and remove sequencing portions of low quality.

There are several tools available for sequence QC. Below we will use one
such tool called
[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/),
which provides a visual illustration of base quality, the distribution
of GC content and nucleotide bias. Similarly, there are multiple tools
suitable to remove low-quality bases, adapter sequences, and other
technical sequences from the reads to improve the accuracy of downstream
analysis. Below we will use a tool called
[Trimmomatic](http://www.usadellab.org/cms/index.php?page=trimmomatic)
to do so, after which we will again assess the quality change with
FastQC. Lastly, to create a comprehensive overview of the quality of ALL
the sequenced samples, we will use [MultiQC,](https://multiqc.info)
which essentially just merged the outputs from your FastQC runs into a
cohesive report.

BUT, before we can run any of these software tools, we need to configure
the appropriate environments and download the correct tools/versions
etc. onto the cluster. So let's begin with that ...

(first be sure to go through the [basics of the
cluster](https://github.com/UC-Boulder/metagenomics_data_processing/blob/main/3.computer_cluster/1.start_here_for_alpine.md))

## 1. Create your environment 

Log into the cluster, get onto a compute node and cd to your scratch
alpine

``` bash
$ ssh emye7956@login.rc.colorado.edu
$ acompile -n 4 --time=2:30:00
$ module load slurm/alpine
$ cd /scratch/alpine/emye7956/
```

activate conda, create the environment. You can set this up manually
using the install instructions for FastQC, Trimmomatic and Multiqc ...
or you can use the yaml file in this directory called qc_env_alpine.yaml
that should set this all up for you.

``` bash
$ module load anaconda
  $ conda env create --file=qc_env_alpine.yaml
```

This will usually print out the path where the environment is can be
found. Use that path to activate the environment. Eg:

``` bash
$ conda activate /scratch/alpine/$USER/software/anaconda/envs/qc_env
```
